### PR TITLE
[CAMEL-15757] Add rawPayload support in camel-salesforce composite

### PR DESCRIPTION
**Pull request by jbonofre**
- [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
- [x] Each commit in the pull request should have a meaningful subject line and body.
- [x] If you're unsure, you can format the pull request title like `[CAMEL-XXX] Fixes bug in camel-file component`, where you replace `CAMEL-XXX` with the appropriate JIRA issue.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Run `mvn clean install -Psourcecheck` in your module with source check enabled to make sure basic checks pass and there are no checkstyle violations. A more thorough check will be performed on your pull request automatically.
Below are the contribution guidelines:
https://github.com/apache/camel/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by jbonofre** [comment]
This PR adds support of Salesforce Composite API. I'm adding a test to show the use case.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/JsonRestProcessor.java [hunk 1]
```diff
@@ -98,6 +98,14 @@ protected void processRequest(Exchange exchange) throws SalesforceException {
                 exchange.setProperty(RESPONSE_CLASS, CreateSObjectResult.class);
                 break;
 
+            /**
+             * TODO AHMEDBS
```
Comment (oscerd) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/JsonRestProcessor.java:102]: Please remove todos

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/XmlRestProcessor.java [hunk 2]
```diff
@@ -124,6 +124,13 @@ protected void processRequest(Exchange exchange) throws SalesforceException {
                 exchange.setProperty(RESPONSE_CLASS, CreateSObjectResult.class);
                 break;
 
+            /**
+             * TODO AHMEDBS
```
Comment (oscerd) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/XmlRestProcessor.java:128]: Ditto

**Review by oscerd — CHANGES_REQUESTED** [review]
Cc @zregvart

**Review by zregvart — CHANGES_REQUESTED** [review]
I could be missing something here, we already have composite support, implemented in [CAMEL-11995](https://issues.apache.org/jira/browse/CAMEL-11995), see [the docs](https://camel.apache.org/components/latest/salesforce-component.html#_using_salesforce_composite_api_to_submit_sobject_tree).

**Comment by zregvart** [comment]
@jbonofre have a look at the [integration test](https://github.com/apache/camel/blob/68161be9829f3eb277e5b6fcf51af298d678b045/components/camel-salesforce/camel-salesforce-component/src/test/java/org/apache/camel/component/salesforce/CompositeApiIntegrationTest.java) we have for the composite API support.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/XmlRestProcessor.java (hunk 2 above)
Comment (jbonofre) [reply components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/XmlRestProcessor.java:128]: I will fix. Thanks.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/JsonRestProcessor.java (hunk 1 above)
Comment (jbonofre) [reply components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/JsonRestProcessor.java:102]: I will. Thanks.

**Comment by oscerd** [comment]
But we already have composite support, so I don't think we need this PR.

**Comment by jbonofre** [comment]
Let me check, the composite API use was different and not same impl.

**Comment by oscerd** [comment]
I'd avoid confusion. We already have an implementation.

**Comment by jbonofre** [comment]
I confirm, it's completely different API, not opposite but complement:

https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_composite.htm

https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobjects_collections_upsert.htm

**Comment by jbonofre** [comment]
The current "composite" API in the component processes 25 elements with dependencies.
The "composite" API in this PR processes 200 elements of same type without dependency between them.

So, this PR extends the salesforce component with "new" composite API.

**Comment by jbonofre** [comment]
@oscerd @zregvart as explained it's now the same APIs. This PR brings the "new" composite API. As this PR doesn't break anything to the existing, I think it's worth to support it (and not confusing for user, I can improve the documentation anyway).

**Comment by oscerd** [comment]
They are great at naming things :-)

**Comment by zregvart** [comment]
TLDR, I think this should be done within the existing composite API code. If it remains as is, we will have to maintain two different composite API implementations and two different user experiences.

What we have currently is support for:

- `composite` (https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_composite.htm)
- `composite-tree` (https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobject_tree.htm)
- `composite-batch` (https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_batch.htm)

This PR adds support for:
- upsert functionality of composite collections (https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobjects_collections_upsert.htm)

I think we need to be consistent with existing implementation to help minimize the maintenance effort, and consistent with the user experience. To that end:

 - I'd like to see this incorporated within the composite support we already have: [CompositeApiProcessor](https://github.com/apache/camel/blob/21a3bb6764dba20698693344b1d3caa901d2f8f2/components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/processor/CompositeApiProcessor.java), [CompositeApiClient](https://github.com/apache/camel/blob/21a3bb6764dba20698693344b1d3caa901d2f8f2/components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/client/CompositeApiClient.java) and [DefaultCompositeApiClient](https://github.com/apache/camel/blob/21a3bb6764dba20698693344b1d3caa901d2f8f2/components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/client/DefaultCompositeApiClient.java).

 - The operation should be named something like `composite-collections-upsert` or `composite-collections` with auto-detection (if possible) of sub-operation (create, retreive, update, upsert, delete). 

 - I'd also like to see full support for composite collections (individually multiple contributions).

**Comment by zregvart** [comment]
Within the realm of composite API, Camel is currently missing:
 - composite graphs added in v50 (https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_graph_introduction.htm)
 - composite collections added in v42 (https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobjects_collections.htm)

**Comment by jbonofre** [comment]
I'm updating the PR to use better API naming. I will also rebase (to solve conflicts).

**Comment by jbonofre** [comment]
I renamed the operation to "CompositeBatch" to avoid the confusion. Thoughts ?

**Comment by zregvart** [comment]
> I renamed the operation to "CompositeBatch" to avoid the confusion. Thoughts ?

Please read the [comment I left earlier](https://github.com/apache/camel/pull/4514#issuecomment-716532649), this adds "upsert functionality of composite collections", we already support composite batch with the operation named `composite-batch`, if anything this adds more confusion.

**Comment by jbonofre** [comment]
Ah my bad, let me change to "upsert_composite_collections" than ;)

**Comment by jbonofre** [comment]
I updated the PR to use `upsertCompositeSObjects` operation name (to avoid confusion).

**Review by oscerd — APPROVED** [review]

**Comment by jbonofre** [comment]
@oscerd thanks ! @zregvart thoughts ?

If you are OK, I will merge this.

**Comment by jbonofre** [comment]
By the way, I will remove from the PR the files generated by the maven build.

**Comment by jbonofre** [comment]
Updated PR after discussion with @zregvart and @jeremyross . I will add `rawPayload` support in other composite operation.

**Comment by jbonofre** [comment]
Just updated the test (composite only supports json format (not XML)).

The PR is good to review: `rawPayload` actually makes sense only for `composite` endpoint (as we can do tree or batch directly with `composite`, the user can prepare the body as he wants).
 
@zregvart @oscerd @jeremyross mind to take a look ?

**Review by oscerd — APPROVED** [review]

**Review by zregvart — APPROVED** [review]
Perfect. Thanks! Would be good to also provide some documentation, I usually just copy-paste and tidy up the code from the integration test in the asciidoc.

**Comment by jbonofre** [comment]
@zregvart true ! let me do that ;)

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/client/DefaultCompositeApiClient.java [hunk 3]
```diff
@@ -93,6 +93,35 @@ public DefaultCompositeApiClient(final SalesforceEndpointConfig configuration, f
...
+    public void submitCompositeRaw(
+            final InputStream raw, final Map<String, List<String>> headers,
+            final ResponseCallback<InputStream> callback,
+            final String sObjectName, final String extId, String compositeMethod)
+            throws SalesforceException {
+        checkCompositeFormat(format, SObjectComposite.REQUIRED_PAYLOAD_FORMAT);
+
+        final String url = String.format("%s%s/%s/%s/%s", versionUrl(), "composite", "sobjects", sObjectName, extId);
+
+        System.out.println("URL: " + url);
```
Comment (davsclaus) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/client/DefaultCompositeApiClient.java:105]: Remove system out, or change into a debug LOG

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java [hunk 4]
```diff
@@ -126,6 +127,8 @@
     @UriParam
     private String apexMethod;
     @UriParam
+    private String compositeMethod;
```
Comment (davsclaus) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java:130]: Can this be used both on consumer or producer? If not then add label to mark it as consumer or producer only.

**Review by davsclaus — CHANGES_REQUESTED** [review]
As others mentions needs some documentation what this compositeMethod does. Also if its consumer or producer only or both.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/client/DefaultCompositeApiClient.java (hunk 3 above)
Comment (jbonofre) [reply components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/internal/client/DefaultCompositeApiClient.java:105]: +1 (was debugging)

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java (hunk 4 above)
Comment (jbonofre) [reply components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java:130]: Like `apexMethod` it's consumer only for now. Let me improve this (we should update `apexMethod` as well but it's another PR).

**Comment by jbonofre** [comment]
I added a documentation. For `compositeMethod`, it's similar to `apexMethod`. Should I mention producer ?

**Review by davsclaus — APPROVED** [review]

**Comment by davsclaus** [comment]
You actually mark on the @UriParam if its producer or consumer only, by setting label = "producer", label = "consumer" etc.

**Review by oscerd — APPROVED** [review]

**Comment by jbonofre** [comment]
@davsclaus yeah, I know. My point is: should I do it only for `compositeMethod` or also for `apexMethod` ? I guess only `compositeMethod` ?

**Comment by davsclaus** [comment]
You are welcome to do it for the options which you in fact know are only either one. So if apex is only producer then you can mark it too.

**Comment by jbonofre** [comment]
Updated, thanks !

**Review by davsclaus — APPROVED** [review]
